adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,123 @@
+
+== Proposal
+
+Moving metadata generation to the 'right' place is the architecturally correct thing to do.
+
+image::images/0002-resolution.svg[the proposed resolution]
+Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg
+
+So how do we minimise the impact?
+What needs to happen for it to work?
+
+* The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module. An alternate implementation would be to publish a separate `-metadata` artifact and the tooling could read that artifact.
+* The build of extensions' deployment module invokes a maven plugin (the existing one, or a new one)
+* (optional) Extension source code refactoring:  the template `runtime/src/META-INF/quarkus-extension.yaml` moves to `deployment`. This eliminates one of the cross-module dependencies we'd like to avoid to stay cache-friendly. We want this refactoring to be optional, so if the yaml template isn't found in the deployment module, tooling would look for it in the runtime module.
+
+The `quarkus-extension.properties` will continue to be generated in the runtime module. We should do an inventory to double-check that nothing in the `quarkus-extension.yaml` needs
+
+* to be accessed by tools outside the registry which expects it to be in the runtime JAR (IDEs?)
+* generated information which is only available to the extension plugin when it's running in the runtime module
+
+If so, the consumers will need to be updated to read the deployment module, or those informations should move to `quarkus-extension.properties` (if the consumer has no access to the deployment module).
+
+A secondary benefit of this proposal is reducing the size (slightly) of the extension's runtime artifacts.
+
+=== Transition plan
+
+We will need to allow a long transition period (probably one major release lifespan), in which
+
+* The registry looks in both the runtime and deployment jars for `META-INF/quarkus-extension.yaml`, and possibly also the new metadata artifact
+* The extension plugin tolerates being invoked from both `runtime` and `deployment` modules (so it's probably simplest to have one 'umbrella' plugin which handles both modules). When invoked, it will look in both the runtime and deployment modules for `src/META-INF/quarkus-extension.yaml`.
+* We still package up a (less complete) `quarkus-extension.yaml` into the runtime module, so that back-level tools (such as the CLI, and maybe IDEs) can find it
+
+image::images/0002-transition-flow.svg[a transitional process, with fallbacks]
+Source::https://excalidraw.com/#json=A7UUcn3Vme98So4xp3jZ6,bemZxMtdn5vLP9MgpZtwJw
+
+In other words, the deployment module becomes the new 'best source', but
```

Comment:
the way I thought about it when making extension.yaml early on was that it would be *merged* with what is in deployment.

so instead of expecting the runtime file to have everything - it is expected you read and combine/merge both.

Reply:
I can see the argument here. The counter-argument is that since the deployment artifact can see what’s in the runtime artifact, why not do the merging in our own mojo, rather than expect all consuming tooling to do the merging? Doing the merging in the mojo is more or less what’s being proposed by this ADR (modulo some discussion about where the source file lives).